### PR TITLE
chore: refactor/centralise secretKeyRef usage

### PR DESCRIPTION
**Pull request by moolen**
This PR standardizes and centralizes the secret fetching process. This is to avoid having too much duplication across providers and to secure it.

### CODE REVIEW

**Comment by Skarlso** [comment]
Ooo. Is that sonar issue new?

**Comment by moolen** [comment]
fixed the duplication issues, i'll tackle the cognitive complexity one separately, i want to restructure & refactor the vault provider. It's not maintainable :laughing:  🫠

File: pkg/generator/gcr/gcr.go [hunk 1]
```diff
@@ -65,7 +65,7 @@ func (g *Generator) generate(
 	ts, err := tokenSource(ctx, esv1beta1.GCPSMAuth{
 		SecretRef:        (*esv1beta1.GCPSMAuthSecretRef)(res.Spec.Auth.SecretRef),
 		WorkloadIdentity: (*esv1beta1.GCPWorkloadIdentity)(res.Spec.Auth.WorkloadIdentity),
-	}, res.Spec.ProjectID, false, kube, namespace)
+	}, res.Spec.ProjectID, "", kube, namespace)
```
Comment (Skarlso) [inline pkg/generator/gcr/gcr.go:68]: Hum, not sure if I like an empty string better than a false. :D

Comment (Skarlso) [reply pkg/generator/gcr/gcr.go:68]: What I'm saying is that this doesn't signify anything when reading. I would have to open the function to figure out what this is. Can you please create a constant saying EmptyStoreKind or something like that? That would make more sense when setting/reading an unused parameter. WDYT?

File: pkg/provider/akeyless/akeyless.go [hunk 2]
```diff
@@ -473,12 +475,12 @@ func (a *akeylessBase) getCertFromSecret(provider *esv1beta1.AkeylessProvider) (
 	}
 
 	ctx := context.Background()
-	res, err := a.secretKeyRef(ctx, &secretRef)
+	cert, err := utils.ResolveSecretKeyRef(ctx, a.kube, a.storeKind, a.namespace, &secretRef)
```
Comment (Skarlso) [inline pkg/provider/akeyless/akeyless.go:478]: The problem with the utils package is that I wouldn't know where to look for a SecretKey resolver. :)

How about, you can put it into the utils folder but create it's own package? Something like `resolvers`?

File: pkg/provider/alibaba/kms.go [hunk 3]
```diff
@@ -39,7 +38,8 @@ const (
 	errUninitalizedAlibabaProvider             = "provider Alibaba is not initialized"
 	errInvalidClusterStoreMissingAKIDNamespace = "invalid ClusterStore, missing  AccessKeyID namespace"
 	errInvalidClusterStoreMissingSKNamespace   = "invalid ClusterStore, missing namespace"
-	errFetchAKIDSecret                         = "could not fetch AccessKeyID secret: %w"
```
Comment (Skarlso) [inline pkg/provider/alibaba/kms.go:42]: While you're here... can you please remove the unused errors? 

```go
	errMissingSAK                              = "missing AccessSecretKey"
	errMissingAKID                             = "missing AccessKeyID"
	errInvalidClusterStoreMissingAKIDNamespace = "invalid ClusterStore, missing  AccessKeyID namespace"
	errInvalidClusterStoreMissingSKNamespace   = "invalid ClusterStore, missing namespace"
        errAlibabaCredSecretName                   = "invalid Alibaba SecretStore resource: missing Alibaba APIKey"
```

File: pkg/utils/secret_ref.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	esmeta "github.com/external-secrets/external-secrets/apis/meta/v1"
+)
+
+const (
+	errGetKubeSecret         = "cannot get Kubernetes secret %q: %w"
+	errSecretKeyFmt          = "cannot find secret data for key: %q"
+	errGetKubeSATokenRequest = "cannot request Kubernetes service account token for service account %q: %w"
+)
+
+// Resolves a metav1.SecretKeySelector and returns the value of the secret it points to.
```
Comment (Skarlso) [inline pkg/utils/secret_ref.go:40]:
```suggestion
// ResolveSecretKeyRef resolves a metav1.SecretKeySelector and returns the value of the secret it points to.
```

File: pkg/utils/secret_ref.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	}
+	tokenResponse, err := corev1Client.ServiceAccounts(tokenRequest.Namespace).
+		CreateToken(ctx, serviceAccountRef.Name, tokenRequest, metav1.CreateOptions{})
+	if err != nil {
+		return "", fmt.Errorf(errGetKubeSATokenRequest, serviceAccountRef.Name, err)
+	}
+	return tokenResponse.Status.Token, nil
+}
+
+func NewKubeClient() (*kubernetes.Clientset, error) {
```
Comment (Skarlso) [inline pkg/utils/secret_ref.go:103]: This function is unused.

**Review by Skarlso — APPROVED** [review]
LGTM with a few suggests and findings. :) Well done! :)))

**Comment by moolen** [comment]
Thank you @Skarlso for your review and the good suggestions, much appreciated! :bow:

File: pkg/generator/gcr/gcr.go [hunk 6]
```diff
@@ -65,7 +66,7 @@ func (g *Generator) generate(
 	ts, err := tokenSource(ctx, esv1beta1.GCPSMAuth{
 		SecretRef:        (*esv1beta1.GCPSMAuthSecretRef)(res.Spec.Auth.SecretRef),
 		WorkloadIdentity: (*esv1beta1.GCPWorkloadIdentity)(res.Spec.Auth.WorkloadIdentity),
-	}, res.Spec.ProjectID, false, kube, namespace)
+	}, res.Spec.ProjectID, resolvers.EmptyStoreKind, kube, namespace)
```
Comment (Skarlso) [inline pkg/generator/gcr/gcr.go:69]: ❤️❤️

File: pkg/utils/resolvers/secret_ref.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+)
+
+const (
+
+	// This is used to determine if a store is cluster-scoped or not.
+	// The EmptyStoreKind is not cluster-scoped, hence resources
+	// cannot be resolved across namespaces.
+	// TODO: when we implement cluster-scoped generators
+	// we can remove this and replace it with a interface.
+	EmptyStoreKind = "EmptyStoreKind"
```
Comment (Skarlso) [inline pkg/utils/resolvers/secret_ref.go:36]: Nice 👌

**Review by Skarlso — APPROVED** [review]
Thanks! Looks fantastic. Nice work doing this. The codebase got a lot better. 😊

**Pull request merged**